### PR TITLE
[enhancement] Registered menu groups #253

### LINKED ISSUE #253: [change] Register menu groups
Must be done after implementing register menu group functionality [openwisp/openwisp-utils#189](https://github.com/openwisp/openwisp-utils/issues/189)

### PR DESCRIPTION
**Pull request by niteshsinha17**
Registered menu groups. This change is needed because ``controller`` is using both ``utils`` and ``users`` but ``users`` is using master branch of ``utils``. I will make changes in the others modules using this PR branch.

- [x] I have read the [contributing guidelines](http://openwisp.io/docs/developer/contributing.html#how-to-commit-your-changes-properly).
- [x] I have manually tested the proposed changes.
- [x] I have written new test cases to avoid regressions. (if necessary)
- [x] I have updated the documentation. (e.g. README.rst)

closes #253

### CODE REVIEW

File: openwisp_users/apps.py [hunk 1]
```diff
@@ -23,16 +23,26 @@ class OpenwispUsersConfig(AppConfig):
...
-        register_menu_items(
-            [
-                {'model': settings.AUTH_USER_MODEL},
-                {'model': get_model_name('openwisp_users', 'Organization')},
-            ]
+    def register_menu_group(self):
+        register_menu_group(
+            position=4,
+            config={
+                'label': 'User',
```
Comment (nemesisdesign) [inline openwisp_users/apps.py:34]: flag the string as translatable and use plural form please

File: openwisp_users/apps.py [hunk 2]
```diff
@@ -23,16 +23,26 @@ class OpenwispUsersConfig(AppConfig):
...
+                'model': settings.AUTH_USER_MODEL,
+                'name': 'changelist',
+            },
+        )
+        register_menu_group(
+            position=5,
+            config={
+                'label': 'Organizations',
```
Comment (nemesisdesign) [inline openwisp_users/apps.py:42]: flag string as translatable, always do this for strings which are shown in the UI

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
See below.
I think we should wait to merge this until this is ready: https://github.com/openwisp/openwisp-utils/pull/185.
Then we should add icons too.

**Comment by niteshsinha17** [comment]
@nemesisdesign We should add ``new-menu`` branch in ci like we did in utils.

**Comment by nemesisdesign** [comment]
> @nemesisdesign We should add `new-menu` branch in ci like we did in utils.

Done.

File: requirements.txt [hunk 3]
```diff
@@ -4,7 +4,7 @@ django-allauth~=0.44.0
 django-phonenumber-field>=4.0,<5.2
 phonenumbers>=8.11.0,<8.13.0
 # TODO: change this when next point version of openwisp-utils is released
-openwisp-utils[rest] @ https://github.com/openwisp/openwisp-utils/tarball/master
+openwisp-utils[rest] @ https://github.com/niteshsinha17/openwisp-utils/tarball/gsoc/new-navigation-menu-design
```
Comment (atb00ker) [inline requirements.txt:7]: TODO: Needs to point to `openwisp` before merge. :smile:

File: openwisp_users/apps.py [hunk 4]
```diff
@@ -23,16 +23,28 @@ class OpenwispUsersConfig(AppConfig):
...
         self.set_default_settings()
         self.connect_receivers()
 
-    def register_default_menu_items(self):
-        register_menu_items(
-            [
-                {'model': settings.AUTH_USER_MODEL},
-                {'model': get_model_name('openwisp_users', 'Organization')},
-            ]
+    def register_menu_group(self):
```
Comment (atb00ker) [inline openwisp_users/apps.py:30]: Code wise it looks good, but try to run the app.
It didn't show Users and Organizations in my menu! :-(

Comment (niteshsinha17) [reply openwisp_users/apps.py:30]: Icons were not added. I will add them.

**Review by atb00ker — COMMENTED** [review]
Seems like home is registered twice: 

e1

**Comment by niteshsinha17** [comment]
> Seems like home is registered twice:
> 
> e1

@atb00ker I have added ``home`` again in the utils to test the menu. It helps in knowing what is the last item in the menu and is it properly visible when we scroll down. It's not an issue. It will be removed from there before merging.

**Review by atb00ker — DISMISSED** [review]

File: tests/testapp/tests/test_filter_classes.py [hunk 5]
```diff
@@ -63,11 +61,9 @@ def test_browsable_api_filter_owner(self):
             user=operator, is_admin=True, organization=self._get_org('org_a')
         )
         token = self._obtain_auth_token()
-        auth = dict(HTTP_AUTHORIZATION=f'Bearer {token}')
+        url = f'{reverse("test_books_list_manager_view", args=(self.shelf_a.id,))}'
```
Comment (atb00ker) [inline tests/testapp/tests/test_filter_classes.py:64]: Hey, why was this changed?

Comment (niteshsinha17) [reply tests/testapp/tests/test_filter_classes.py:64]: I rebased this branch last time. I will remove this commit for here.

Comment (atb00ker) [reply tests/testapp/tests/test_filter_classes.py]: Okay, thanks

**Comment by atb00ker** [comment]
Here is what I am seeing:
paru

**Comment by niteshsinha17** [comment]
It looks like it's happening with you only. See this. I have done a fresh installation to test this.
image

**Comment by kkreitmair** [comment]
I have done a fresh installation of this PR and the menu appeared as expected.
image

I also reviewed the changes.
The PR LGTM.

File: openwisp_users/tests/test_admin.py [hunk 6]
```diff
@@ -1457,6 +1457,24 @@ def test_admin_add_user_with_invalid_email(self):
...
+        for model in models:
+            with self.subTest(f'test_admin_group_for_{model}_model'):
+                url = reverse(f'admin:{self.app_label}_{model}_changelist')
+                self.assertContains(response, f'<a class="mg-link" href="{url}">')
+        with self.subTest('test_user_and_organization_group_is_registered'):
+            self.assertContains(
+                response,
+                '<div class="mg-dropdown-label">Users & Organizations </div>',
+                html=True,
+            )
```
Comment (nemesisdesign) [inline openwisp_users/tests/test_admin.py:1476]: please assert the presence of auth.Group

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
auth.Group is missing.

Should be:

- Users & Organizations
  - Users
  - Organizations
  - Organization Owners
  - Organization Users
  - Groups & Permissions  (I suggest putting the "permissions" word here just to make it clear that it deals with permissions)

Regarding icons, here's some suggestions.

The [main icon could be one of these showing on flaticon for organization](https://www.flaticon.com/search?word=organization&license=selection&style=all&order_by=4&type=icon).
Eg:
user
The current icon for users & organizations looks a bit out of context because of that roof.

The icon for the users page is good but it could be just the plain user icon insead of this one which has some kind of circle on its bottom right corner and I can't see very well what that cirlce means.

For [permission groups we could use something like this one from flaticons](https://www.flaticon.com/search?word=group&type=icon&license=selection&order_by=4).

group (1)

The ther icons are fine.

File: openwisp_users/accounts/templates/account/base.html [hunk 7]
```diff
@@ -1 +1 @@
-{% extends "base.html" %}
+{% extends "admin/base.html" %}
```
[the remark this answers is not in the document]
Comment (niteshsinha17) [reply openwisp_users/accounts/templates/account/base.html:1]: I think we don't need to change it

[the remark this answers is not in the document]
Comment (niteshsinha17) [reply openwisp_users/accounts/templates/account/base.html:1]: Without this. I was getting ``Template Does not exists`` during test

**Comment by niteshsinha17** [comment]
@nemesisdesign Thanks for your explaination

**Review by nemesisdesign — APPROVED** [review]
